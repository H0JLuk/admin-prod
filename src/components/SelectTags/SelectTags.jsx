import React from 'react';
import { CloseOutlined, DownOutlined } from '@ant-design/icons';
import { Checkbox, Select } from 'antd';

import styles from './SelectTags.module.css';

function SelectTags({
    showClearIcon = true,
    canRemoveSelected = true,
    onChange,
    data = [],
    value = [],
    nameKey = 'name',
    idKey = 'code',
    placeholder,
}) {
    const stringValue = value.map(String);

    const onRemoveSelectedTag = () => {
        onChange([]);
    };

    const selectTagRender = ({ value, onClose }) => (
        value && (
            <div className={ styles.tagSelect } >
                <p className={ styles.selectText }>{ findName(data, value, nameKey, idKey) }</p>
                { canRemoveSelected && (
                    <CloseOutlined
                        height="15px"
                        width="15px"
                        fill="#09A552"
                        onClick={ onClose }
                    />
                ) }
            </div>
        )
    );

    const options = data.map((elem) => {
        const name = String(elem[nameKey]);
        const value = String(elem[idKey]);

        return {
            label: (
                <>
                    <Checkbox
                        className={ styles.mrr }
                        checked={ stringValue.includes(value) }
                    />
                    { name }
                </>
            ),
            value,
            disabled: elem.disabled,
        };
    });

    return (
        <Select
            className={ styles.select }
            suffixIcon={ () => suffixBlock(stringValue.length, onRemoveSelectedTag, showClearIcon) }
            showArrow
            showSearch={ false }
            maxTagCount="responsive"
            maxTagTextLength={ 50 }
            mode="tags"
            placeholder={ placeholder }
            maxTagPlaceholder={ <span>...</span> }
            tagRender={ selectTagRender }
            onChange={ onChange }
            dropdownClassName={ styles.dropdown }
            options={ options }
            value={ stringValue }
        />
    );
}

function suffixBlock(selectedCount, onRemoveSelectedTag, showClearIcon) {
    return (
        <div className={ styles.suffixBlock }>
            <DownOutlined className={ styles.icon } />
            <div className={ styles.closeIcon }>
                { !!selectedCount && <div className={ styles.selectedCount }>{ selectedCount }</div> }
                { showClearIcon && !!selectedCount && (
                    <CloseOutlined
                        className={ styles.icon }
                        size={ 25 }
                        onClick={ onRemoveSelectedTag }
                    />
                ) }
            </div>
        </div>
    );
}

function findName(data, selectedValue, nameKey, keyToCompare) {
    const name = data.find((elem) => String(elem[keyToCompare]) === selectedValue);

    return name && name[nameKey];
}

export default SelectTags;
