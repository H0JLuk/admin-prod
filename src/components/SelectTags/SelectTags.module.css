/* select */
.select {
    background-color: #fff;
    border: 1px solid #d9d9d9;
    display: flex;
    height: 40px;
    transition: border 0.3s;
}

.select :global(.ant-select-selector) {
    flex-wrap: nowrap;
    overflow: hidden;
    background-color: transparent;
    border: none;
    width: 100%;
}

.select :global(.ant-select-arrow .anticon:not(.ant-select-suffix):not([aria-label='close'])) {
    pointer-events: none;
    transition: 0.3s;
}

.select:global(.ant-select-open .ant-select-arrow) {
    transform: none;
}

.select:global(.ant-select-open .ant-select-arrow .anticon-down:not(.ant-select-suffix)) {
    transform: rotate(-180deg);
}

.select :global(.ant-select-selection-item) {
    background-color: transparent;
    border: none;
}

.select :global(.ant-select-selection-search-input) {
    display: none;
}

.select :global(.ant-select-selection-overflow) {
    max-width: 85%;
}

.select :global(.ant-select-selection-overflow) {
    max-width: 85%;
}


.select :global(.ant-select-selection-search-input) {
    display: none;
}

.select:global(.ant-select:not(.ant-select-customize-input) .ant-select-selector) {
    border: none;
}

/*  /select */

/* tagSelect */
.tagSelect {
    padding: 4px 10px;
    background-color: rgba(9, 165, 82, 0.08);
    color: #09a552;
    border-radius: 25px;
    display: flex;
    align-items: center;
    margin-left: 5px;
}

.tagSelect:global(.ant-select-selection-overflow-item) {
    margin-left: 5px;
}

.tagSelect:global(.ant-select-selection-overflow-item:first-child) {
    margin: 0;
}

.tagSelect > p {
    margin-right: 3px;
}

.tagSelect:global(.ant-select-selection-overflow-item) {
    margin-left: 5px;
}

.tagSelect:global(.ant-select-selection-overflow-item:first-child) {
    margin: 0;
}
/* /tagSelect */

/* dropdown */
.dropdown :global(.ant-select-item-option-selected:not(.ant-select-item-option-disabled)) {
    background: transparent;
    background-color: transparent;
}

.dropdown :global(.ant-select-item-option-state) {
    display: none;
}
/* /dropdown */

.selectText {
    max-width: 120px;
    white-space: nowrap;
    text-overflow: ellipsis;
    overflow: hidden;
}

.arrowActive {
    transform: rotate(180deg);
}

.mrr {
    margin-right: 10px;
}

.icon {
    margin: 0 5px;
    font-size: 15px;
}
.closeIcon {
    display: flex;
    align-items: center;
}

.suffixBlock {
    display: flex;
    height: 100%;
    align-items: center;
    flex-direction: row-reverse;
}

.selectedCount {
    display: flex;
    align-items: center;
    justify-content: center;
    background-color: rgba(9, 165, 82, 0.08);
    margin-left: 10px;
    text-align: center;
    width: 40px;
    height: 28px;
    border-radius: 16px;
    background-color: rgba(9, 165, 82, 0.08);
    margin-left: 10px;
    text-align: center;
    border-radius: 40%;
    color: #09a552;
}
