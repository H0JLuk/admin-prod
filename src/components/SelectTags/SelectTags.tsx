import React from 'react';
import { CloseOutlined, DownOutlined } from '@ant-design/icons';
import { Checkbox, Select } from 'antd';
import { SelectProps } from 'antd/es/select';

import styles from './SelectTags.module.css';

interface ISelectTags {
    showClearIcon?: boolean;
    canRemoveSelected?: boolean;
    onChange?: (value: string[]) => void;
    data: SelectTagsOptions[];
    value?: number[];
    nameKey?: string;
    idKey?: string;
    placeholder: string;
}

// TODO: Придумать как обобщить тип options с помощью дженерика
type SelectTagsOptions = Record<string, any>;

const SelectTags: React.FC<ISelectTags> = ({
    showClearIcon = true,
    canRemoveSelected = true,
    onChange,
    data = [],
    value = [],
    nameKey = 'name',
    idKey = 'code',
    placeholder,
}) => {
    const stringValue = value.map(String);

    const onRemoveSelectedTags = () => {
        onChange && onChange([]);
    };

    const selectTagRender: SelectProps<string[]>['tagRender'] = ({ value, onClose }) => (
        value ? (
            <div className={styles.tagSelect} >
                <p className={styles.selectText}>{findName(data, value as string, nameKey, idKey)}</p>
                {canRemoveSelected && (
                    <CloseOutlined
                        height="15px"
                        width="15px"
                        onClick={onClose}
                    />
                )}
            </div>
        ) : <span />
    );

    const options = data.map((elem) => {
        const name = String(elem[nameKey]);
        const value = String(elem[idKey]);
        return {
            label: (
                <>
                    <Checkbox
                        className={styles.mrr}
                        checked={stringValue.includes(value)}
                    />
                    {name}
                </>
            ),
            value,
            disabled: elem.disabled,
        };
    });

    return (
        <Select<string[]>
            className={styles.select}
            suffixIcon={() => suffixBlock(stringValue.length, onRemoveSelectedTags, showClearIcon)}
            showArrow
            showSearch={false}
            maxTagCount="responsive"
            maxTagTextLength={50}
            mode="tags"
            placeholder={placeholder}
            maxTagPlaceholder={<span>...</span>}
            tagRender={selectTagRender}
            onChange={onChange}
            dropdownClassName={styles.dropdown}
            options={options}
            value={stringValue}
        />
    );
};

function suffixBlock(selectedCount: number, onRemoveSelectedTag: () => void, showClearIcon?: boolean) {
    return (
        <div className={styles.suffixBlock}>
            <DownOutlined className={styles.icon} />
            <div className={styles.closeIcon}>
                {!!selectedCount && <div className={styles.selectedCount}>{selectedCount}</div>}
                {showClearIcon && !!selectedCount && (
                    <CloseOutlined
                        className={styles.icon}
                        size={25}
                        onClick={onRemoveSelectedTag}
                    />
                )}
            </div>
        </div>
    );
}

function findName(data: SelectTagsOptions[], selectedValue: string, nameKey: string, keyToCompare: string) {
    const name = data.find((elem) => String(elem[keyToCompare]) === selectedValue);

    return name && String(name[nameKey]);
}

export default SelectTags;
