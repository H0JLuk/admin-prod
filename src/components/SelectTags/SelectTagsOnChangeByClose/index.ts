import SelectTagsOnChangeByClose from './SelectTagsOnChangeByClose';

export default SelectTagsOnChangeByClose;
