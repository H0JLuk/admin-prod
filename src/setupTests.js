import '@testing-library/jest-dom';
import { configure as enzymeConfigure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { act } from 'react-dom/test-utils';
import './jestMocks';

enzymeConfigure({ adapter: new Adapter() });

const globalTimeout = global.setTimeout;

export const sleep = async (timeout = 0) => {
    await act(async () => {
        await new Promise((resolve) => globalTimeout(resolve, timeout));
    });
};
