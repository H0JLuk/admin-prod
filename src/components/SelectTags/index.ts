import SelectTags from './SelectTags';

export type { ISelectTags } from './SelectTags';

export default SelectTags;
