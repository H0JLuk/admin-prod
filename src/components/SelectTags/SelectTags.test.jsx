import React from 'react';
import SelectTags from './SelectTags';
import { mount, shallow } from 'enzyme';
import { act, fireEvent, render } from '@testing-library/react';
import { sleep } from '../../setupTests';

jest.mock('react', () => ({
    ...jest.requireActual('react'),
    useLayoutEffect: jest.requireActual('react').useEffect,
}));

describe('<SelectTags /> tests', () => {
    const props = {
        onChange: jest.fn(),
        data: [
            {
                active: true,
                categoryDescription: '-50%',
                categoryId: 8,
                categoryName: 'Организуем приём у лучших врачей онлайн 24/7',
                categoryUrl: 'http://distributor-fs:8081/distributor-fs/file?path=category/6.png',
                highlights: [],
            },
        ],
        value: ['11', '8'],
        nameKey: 'categoryName',
        idKey: 'categoryId',
        placeholder: 'Выберите категорию',
    };

    it('SelectTags snapshot', () => {
        const SelectTagsItem = shallow(<SelectTags { ...props } />);
        expect(SelectTagsItem.html()).toMatchSnapshot();
    });

    it('should clear select and call onChange func', () => {
        const { container } = render(<SelectTags { ...props } />);
        fireEvent.click(container.getElementsByClassName('anticon-close')[0]);
        expect(props.onChange).toBeCalled();
    });

    it('should open dropdown and call onChange func', async () => {
        const newProps = { ...props, value: [] };
        const wrapper = render(<SelectTags { ...newProps } />);

        await act(async() => {
            fireEvent.mouseDown(wrapper.getByText(props.placeholder));
            await sleep();
            expect(document.querySelector('.ant-select-dropdown')).toBeInTheDocument();
        });

        await act(async() => {
            fireEvent.click(wrapper.getByText(props.data[0].categoryName));
            await sleep();
            expect(props.onChange).toBeCalled();
        });
    });

    it('should call tagRender func and return tagSelect element', async() => {
        jest.useFakeTimers();
        const onClose = jest.fn();
        const wrapper = mount(<SelectTags { ...props } />);

        wrapper.find('.ant-select-selector').simulate('mousedown');
        jest.runAllTimers();
        wrapper.update();

        await act(async() => {
            const {
                props: { className },
            } = wrapper.find('Select').prop('tagRender')({ value: props.value[1], onClose });
            expect(className).toEqual('tagSelect');
        });
    });

    it('should have str[] in select value prop', () => {
        const newProps = { ...props, value: [11, 8] };
        const SelectTagsItem = shallow(<SelectTags { ...newProps } />);
        expect(SelectTagsItem.prop('value')[0]).toEqual(newProps.value[0].toString());
        expect(SelectTagsItem.prop('value')[1]).toEqual(newProps.value[1].toString());
    });
});
