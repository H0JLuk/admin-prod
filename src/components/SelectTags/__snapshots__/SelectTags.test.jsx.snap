// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectTags /> tests SelectTags snapshot 1`] = `"<div class=\\"ant-select select ant-select-multiple ant-select-show-arrow\\"><div class=\\"ant-select-selector\\"><div class=\\"ant-select-selection-overflow\\"><div class=\\"ant-select-selection-overflow-item ant-select-selection-overflow-item-rest\\" style=\\"opacity:0;height:0;overflow-y:hidden;order:9007199254740991;pointer-events:none;position:absolute\\" aria-hidden=\\"true\\"><span class=\\"ant-select-selection-item\\"><span class=\\"ant-select-selection-item-content\\"><span>...</span></span></span></div><div class=\\"ant-select-selection-overflow-item ant-select-selection-overflow-item-suffix\\" style=\\"opacity:1;order:0\\"><div class=\\"ant-select-selection-search\\" style=\\"width:0\\"><input type=\\"search\\" autoComplete=\\"off\\" class=\\"ant-select-selection-search-input\\" role=\\"combobox\\" aria-haspopup=\\"listbox\\" aria-owns=\\"undefined_list\\" aria-autocomplete=\\"list\\" aria-controls=\\"undefined_list\\" aria-activedescendant=\\"undefined_list_0\\" value=\\"\\"/><span class=\\"ant-select-selection-search-mirror\\" aria-hidden=\\"true\\">Â </span></div></div></div></div><span class=\\"ant-select-arrow\\" style=\\"user-select:none;-webkit-user-select:none\\" unselectable=\\"on\\" aria-hidden=\\"true\\"><div class=\\"suffixBlock\\"><span role=\\"img\\" aria-label=\\"down\\" class=\\"anticon anticon-down icon\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" data-icon=\\"down\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z\\"></path></svg></span><div class=\\"closeIcon\\"><div class=\\"selectedCount\\">2</div><span role=\\"img\\" aria-label=\\"close\\" size=\\"25\\" tabindex=\\"-1\\" class=\\"anticon anticon-close icon\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" data-icon=\\"close\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z\\"></path></svg></span></div></div></span></div>"`;
